File: platform/android/MapLibreAndroid/src/multiBackend/java/org/maplibre/android/module/loader/LibraryLoaderProviderImpl.java
package org.maplibre.android.module.loader;

import org.maplibre.android.LibraryLoader;
import org.maplibre.android.LibraryLoaderProvider;
import org.maplibre.android.RenderingEngine;

/**
 * multiBackend flavor: when asked to load "maplibre", routes to either
 * libmaplibre.so (OpenGL, the default name) or libmaplibre-vulkan.so based on
 * {@link RenderingEngine#getCurrentType()}. Any other library name is loaded
 * verbatim.
 */
public class LibraryLoaderProviderImpl implements LibraryLoaderProvider {

  @Override
  public LibraryLoader getDefaultLibraryLoader() {
    return new MultiBackendLibraryLoader();
  }

  private static class MultiBackendLibraryLoader extends LibraryLoader {
    @Override
    public void load(String name) {
      if ("maplibre".equals(name)
          && RenderingEngine.getCurrentType() == RenderingEngine.Type.VULKAN) {
        System.loadLibrary("maplibre-vulkan");
      } else {
        System.loadLibrary(name);
      }
    }
  }
}

File: platform/android/MapLibreAndroid/src/multiBackend/java/org/maplibre/android/module/loader/LibraryLoaderProviderImpl.kt
package org.maplibre.android.module.loader

import org.maplibre.android.LibraryLoader
import org.maplibre.android.LibraryLoaderProvider
import org.maplibre.android.RenderingEngine

/**
 * multiBackend flavor: when asked to load "maplibre", routes to either
 * libmaplibre.so (Vulkan, the default name) or libmaplibre-opengl.so based on
 * [RenderingEngine.getCurrentType]. Any other library name is loaded
 * verbatim.
 */
class LibraryLoaderProviderImpl : LibraryLoaderProvider {
    override fun getDefaultLibraryLoader(): LibraryLoader = MultiBackendLibraryLoader()

    private class MultiBackendLibraryLoader : LibraryLoader() {
        override fun load(name: String) {
            if ("maplibre" == name && RenderingEngine.currentType == RenderingEngine.Type.OPENGL) {
                System.loadLibrary("maplibre-opengl")
            } else {
                System.loadLibrary(name)
            }
        }
    }
}
